cpp/test/neighbors/ann_vamana.cuh

```diff
@@ -0,0 +1,219 @@
+  EXPECT_TRUE(index_->graph().extent(1) == inputs.graph_degree);
+
+  // Copy graph to host
+  auto h_graph = raft::make_host_matrix<IdxT, int64_t>(inputs.n_rows, inputs.graph_degree);
+  raft::copy(h_graph.data_handle(), index_->graph().data_handle(), index_->graph().size(), stream);
+
+  size_t edge_count = 0;
+  int max_degree    = 0;
+  for (int i = 0; i < h_graph.extent(0); i++) {
+    int temp_degree = 0;
+    for (int j = 0; j < h_graph.extent(1); j++) {
+      if (h_graph(i, j) < (uint32_t)(inputs.n_rows)) temp_degree++;
+    }
+    if (temp_degree > max_degree) max_degree = temp_degree;
+    edge_count += (size_t)temp_degree;
+  }
+
+  // Tests for acceptable range of edges - low dim can also impact this
+  // Minimum expected maximum degree across the whole graph
+  EXPECT_TRUE(max_degree >= std::min(inputs.graph_degree, inputs.dim));
+
+  float max_edges = (float)(inputs.n_rows * std::min(inputs.graph_degree, inputs.dim));
+
+  RAFT_LOG_INFO("dim:%d, degree:%d, visited_size:%d, edge_count:%lu, max_edges:%lu\n",
+                inputs.dim,
+                inputs.graph_degree,
+                inputs.visited_size,
+                edge_count,
+                (size_t)max_edges);
+
+  // Graph won't always be full, but <75% is very unlikely
+  EXPECT_TRUE(((float)edge_count / max_edges) > 0.75);
+
+  // TODO - Anything else we can test without search
+}
+
+template <typename DistanceT, typename DataT, typename IdxT>
+class AnnVamanaTest : public ::testing::TestWithParam<AnnVamanaInputs> {
+ public:
+  AnnVamanaTest()
+    : stream_(raft::resource::get_cuda_stream(handle_)),
+      ps(::testing::TestWithParam<AnnVamanaInputs>::GetParam()),
+      database(0, stream_)
+  {
+  }
+
+ protected:
+  void testVamana()
+  {
+    vamana::index_params index_params;
+    index_params.metric       = ps.metric;
+    index_params.graph_degree = ps.graph_degree;
+    index_params.visited_size = ps.visited_size;
+    index_params.max_fraction = ps.max_fraction;
+
+    auto database_view = raft::make_device_matrix_view<const DataT, int64_t>(
+      (const DataT*)database.data(), ps.n_rows, ps.dim);
+
+    //        {
+    vamana::index<DataT, IdxT> index(handle_);
+    if (ps.host_dataset) {
+      auto database_host = raft::make_host_matrix<DataT, int64_t>(ps.n_rows, ps.dim);
+      raft::copy(database_host.data_handle(), database.data(), database.size(), stream_);
+      auto database_host_view = raft::make_host_matrix_view<const DataT, int64_t>(
+        (const DataT*)database_host.data_handle(), ps.n_rows, ps.dim);
+
+      index = vamana::build(handle_, index_params, database_host_view);
+    } else {
+      index = vamana::build(handle_, index_params, database_view);
+    };
+
+    CheckGraph<DataT, IdxT>(&index, ps, stream_);
+
+    // Can we test serialize here without deserialize implemented?
+    //          cagra::serialize(handle_, "cagra_index", index, ps.include_serialized_dataset);
```

Comment:
I think it would be good to call serialize, even if it is just a smoke test.

Reply:
I considered doing this, but wasn't sure how it impacts CI to actually write files out, etc. Do we do this with other tests?

Reply:
Added this as part of the test.